fuzzers/main.py

```diff
@@ -0,0 +1,337 @@
+FUZZ_CORPUS_DELETED = prometheus_client.Counter('fuzz_corpus_deleted', 'Number of elements deleted from GCS corpus', ['branch', 'crate', 'runner', 'flags'])
+
+def update_repo(branch):
+    if not REPO_DIR.exists():
+        print(f"Doing initial clone of repository {REPO_DIR}")
+        subprocess.check_call(['git', 'clone', REPO_URL, REPO_DIR_NAME], cwd=REPO_DIR_PARENT)
+
+    print(f"Updating to latest commit of branch {branch}")
+    subprocess.check_call(['git', 'fetch', REPO_URL, branch], cwd=REPO_DIR)
+    subprocess.check_call(['git', 'checkout', 'FETCH_HEAD'], cwd=REPO_DIR)
+
+def parse_config():
+    return toml.load(REPO_DIR / 'nightly' / 'fuzz.toml')
+
+FUZZERS = []
+
+def report_artifact(gcs_path, branch, t):
+    # todo: report on zulip with all the details
+    FUZZ_ARTIFACTS_FOUND.labels(branch['name'], t['crate'], t['runner'], t['flags']).inc()
+
+def artifact_deleted(path):
+    raise RuntimeError(f"Deleted artifact {path}, which should have been prevented by ignore_deletions")
+
+def run_fuzzer(branch, t, gcs, exit_event):
+    # First, figure out the current corpus version
+    bucket = gcs.get_bucket(GCS_BUCKET)
+    corpus_vers = bucket.blob('current-corpus').download_as_text().strip()
+
+    date = datetime.datetime.now().strftime('%Y-%m-%d')
+    log_path = pathlib.Path(date) / t['crate'] / t['runner'] / str(uuid.uuid4())
+    log_filename = WORKDIR / 'fuzz-logs' / log_path
+    utils.mkdirs(log_filename.parent)
+    with open(log_filename, 'a') as log_file:
+        current_commit = subprocess.run(['git', 'rev-parse', 'HEAD'], cwd=REPO_DIR, capture_output=True, check=True).stdout
+        log_file.write(f"Corpus version: {corpus_vers}\n")
+        log_file.write(f"Running on commit {current_commit}\n")
+        log_file.flush()
+
+        # Then, make sure the local corpus is the same as the gcs version
+        reset_to_gcs(t['crate'] + '/corpus/' + t['runner'], REPO_DIR, corpus_vers, log_file)
+        reset_to_gcs(t['crate'] + '/artifacts/' + t['runner'], REPO_DIR, corpus_vers, log_file)
+
+        # Then, make sure we auto-upload any new discovery to gcs
+        stop_inotify_event = threading.Event()
+        auto_upload_to_gcs(
+            stop_inotify_event,
+            t['crate'] + '/corpus/' + t['runner'],
+            REPO_DIR,
+            corpus_vers,
+            bucket,
+            log_file,
+            False,
+            lambda path: FUZZ_CORPUS_UPLOADED.labels(branch['name'], t['crate'], t['runner'], t['flags']).inc(),
+            lambda path: FUZZ_CORPUS_DELETED.labels(branch['name'], t['crate'], t['runner'], t['flags']).inc(),
+        )
+        auto_upload_to_gcs(
+            stop_inotify_event,
+            t['crate'] + '/artifacts/' + t['runner'],
+            REPO_DIR,
+            corpus_vers,
+            bucket,
+            log_file,
+            True,
+            lambda path: report_artifact(path, branch, t),
+            artifact_deleted,
+        )
```

Comment:
Yes, that means there's reduced performance because each new corpus item would then be uploaded multiple times, but shouldn't affect correctness because multiple parallel fuzzers could be doing the same anyway. My thoughts were to fix that at the same time as https://github.com/near/nayduck/pull/28#discussion_r799451484 ; would that make sense with you?
(I'm trying to do the minimum required for an initial landing for now because of https://github.com/near/nearcore/pull/6182#issuecomment-1021246546 ; if this can't land without these two changes that would probably take a bit of time, I'll probably grab a gcp vm and get it to start fuzzing while I'd be adjusting the logic to avoid this issue :))

Reply:
I mean, I’m happy to start testing this code on NayDuck machines even before we get the code pushed. You’ll need to ask Sandi to get you access to the Nayduck GCP project and than we can set up the fuzzer first on one machine to see how it behaves.